lib/WeBWorK/PG/ConvertToPGML.pm

```diff
@@ -0,0 +1,210 @@
+# the terms of either: (a) the GNU General Public License as published by the
+# Free Software Foundation; either version 2, or (at your option) any later
+# version, or (b) the "Artistic License" which comes with this package.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.  See either the GNU General Public License or the
+# Artistic License for more details.
+################################################################################
+
+=head1 NAME
+
+WeBWorK::PG::ConvertToPGML
+
+=head1 DESCRIPTION
+
+Converts a pg file to PGML format.
+
+=head1 OPTIONS
+
+=cut
+
+package WeBWorK::PG::ConvertToPGML;
+use parent qw(Exporter);
+
+use strict;
+use warnings;
+
+our @EXPORT = qw(convertToPGML);
+
+# This subroutine converts the file that is passed in as a multi-line string and
+# assumed to be an older-style PG file with BEGIN_TEXT/END_TEXT, BEGIN_SOLUTION/END_SOLUTION,
+# and BEGIN_HINT/END_HINT blocks.
+
+# * parses the loadMacros line(s) to include PGML.pl (and eliminate MathObjects.pl, which)
+#    is imported by PGML.pl.  This also adds PGcourse.pl to the end of the list.
+
+# input is a string containing the source of the pg file to be converted.
+# returns a string that is the converted input string.
+
+sub convertToPGML {
+	my ($pg_source) = @_;
+	my @pgml_block;
+	my $in_pgml_block = 0;
+	my @all_lines;
+
+	my @rows = split(/\n/, $pg_source);
+
+	while (@rows) {
+		my $row = shift @rows;
+		if ($row =~ /BEGIN_(TEXT|HINT|SOLUTION)/ || $row =~ /SOLUTION\(EV3\(<<\'END_SOLUTION\'\)\);/) {
+			push(@pgml_block, $row);
+			$in_pgml_block = 1;
+		} elsif ($row =~ /END_(TEXT|HINT|SOLUTION)/) {
+			push(@pgml_block, $row);
+			$in_pgml_block = 0;
+			push(@all_lines, @{ convertPGMLBlock(\@pgml_block) });
+			@pgml_block = ();
+		} elsif ($in_pgml_block) {
+			push(@pgml_block, $row);
+		} elsif ($row =~ /loadMacros\(/) {
+			# Parse the macros, which may be on multiple rows.
+			# Remove comments within loadMacros block (should we keep them?)
+			my $macros = $row;
+			while ($row && $row !~ /\);\s*$/) {
+				$row = shift @rows;
+				my @mrow = split(/#/, $row);
+				# This only adds the row if there is something relevent to the left of a #
+				$macros .= $mrow[0] if $mrow[0] !~ /^\s*$/;
+			}
+			# Split by commas and pull out the quotes.
+			my @macros =
+				grep { $_ !~ /^#/ }
+				grep {
+					$_ !~ /(PGstandard|PGML|PGauxiliaryFunctions|PGbasicmacros|PGanswermacros|MathObjects|PGcourse).pl/
+				}
+				map {s/['"\s]//gr}
+				split(/\s*,\s*/, $macros =~ s/loadMacros\((.*)\)\;$/$1/r);
+
+			# my @macros = map {s/['"\s]//gr} split(/\s*,\s*/, $macros =~ s/loadMacros\((.*)\)\;$/$1/r);
+			# @macros = grep {
+			# 	$_ !~ /(PGstandard|PGML|PGauxiliaryFunctions|PGbasicmacros|PGanswermacros|MathObjects|PGcourse).pl/
+			# } @macros;
+			# @macros = grep { $_ !~ /^#/ } @macros;
```

Comment:
Delete this commented out code.  It seems to be a remnant of the previous code from the preceding lines.

Reply:
done